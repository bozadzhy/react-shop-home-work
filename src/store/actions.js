import { createAction } from "@reduxjs/toolkit";

import { sendRequest } from "../helpers/sendRequst";

export const actionAddProducts = createAction("ACTION_ADD_PRODUCTS");
export const actionAddToFavotites= createAction("ACTION_ADD_TO_FAVOTITES");
export const actionAddToCart= createAction("ACTION_ADD_TO_CART");
export const actionHandelModal= createAction("ACTION_HANDEL_MODAL");
export const actionMinusFromCart= createAction("ACTION_MINUS_FROM_CART");
export const actionClearCart= createAction("ACTION_CLEAR_CART");
export const actionDeleteItem = createAction("ACTION_DELETE_ITEM");


export const actionFetchProducts = () => (dispatch) => {
  return (sendRequest('products.json').then((results) => {
    dispatch(actionAddProducts(results));
  }))
};

