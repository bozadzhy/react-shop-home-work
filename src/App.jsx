import {useEffect, useContext} from "react";
import { Routes, Route } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";

import ProductList from "./pages/ProductList/ProductList.jsx";
import Cart from "./pages/Cart/Cart.jsx";
import Favorites from "./pages/Favorites/Favorites.jsx";
import NotFound from "./pages/NotFound/NotFound.jsx";
import { actionFetchProducts } from "./store/actions.js";

import "./App.css";
import MainLayout from "./components/MainLayout/MainLayout.jsx";
import { NewThemeContext } from "./context/context.jsx";

function App() {

  const { isTheme } = useContext(NewThemeContext);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(actionFetchProducts());
  }, []);

  const cart = useSelector((state) => state.cart);
  const totalPrice = useSelector((state) => state.totalPrice);

  useEffect(() => {
    localStorage.setItem("cart", JSON.stringify(cart));
    localStorage.setItem("totalPrice", JSON.stringify(totalPrice));
    localStorage.setItem("isTheme", JSON.stringify(isTheme));
  }, [cart, totalPrice, isTheme]);

  return (
    <>
      <Routes>
        <Route path="/" element={<MainLayout />}>
          <Route index element={<ProductList />} />
          <Route path="cart" element={<Cart />} />
          <Route path="favorites/*" element={<Favorites />} />
          <Route path="*" element={<NotFound />} />
        </Route>
      </Routes>
    </>
  );
}

export default App;
