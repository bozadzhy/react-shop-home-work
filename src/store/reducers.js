import { createReducer } from "@reduxjs/toolkit";
import * as actions from "./actions.js";

const initialState = {
  products: [],
  favorites: JSON.parse(localStorage.getItem("favorite")) || [],
  cart: JSON.parse(localStorage.getItem("cart")) || [],
  isModal: false,
  totalPrice: JSON.parse(localStorage.getItem("totalPrice")) || 0,
};

export default createReducer(initialState, {
  [actions.actionAddProducts]: (state, { payload }) => {
    state.products = payload;
  },

  [actions.actionAddToFavotites]: (state, { payload }) => {
    const findObj = state.favorites.find((obj) => obj.id === payload.id);
    if (!findObj || findObj.length === 0) {
      localStorage.setItem(
        "favorite",
        JSON.stringify([...state.favorites, payload])
      );
      state.favorites = [...state.favorites, payload];
    } else {
      const filterFevorites = state.favorites.filter(
        (obj) => obj.id !== payload.id
      );
      localStorage.setItem("favorite", JSON.stringify(filterFevorites));
      state.favorites = [...filterFevorites];
    }
  },

  [actions.actionAddToCart]: (state, { payload }) => {
    const findCart = state.cart.find((obj) => obj.id === payload.id);

    if (!findCart || findCart.length === 0) {
      state.cart = [...state.cart, { ...payload, count: 1 }];

      state.totalPrice = state.cart.reduce((sum, obj) => {
        return obj.price * obj.count + sum;
      }, 0);
    } else {
      findCart.count++;
      state.totalPrice = state.cart.reduce((sum, obj) => {
        return obj.price * obj.count + sum;
      }, 0);
    }
  },

  [actions.actionMinusFromCart]: (state, { payload }) => {
    const findItem = state.cart.find((obj) => obj.id === payload.id);
    findItem.count--;
    if (findItem.count < 1) {
      findItem.count = 1;
    }
    state.totalPrice = state.totalPrice - findItem.price;
  },

  [actions.actionDeleteItem]: (state, { payload }) => {
    const findItem = state.cart.find((obj) => obj.id === payload.id);

    state.totalPrice = state.totalPrice - findItem.price * findItem.count;
    findItem.count--;

    const filterArr = state.cart.filter((obj) => obj.id !== payload.id);
    localStorage.setItem("cart", JSON.stringify(filterArr));
    state.cart = [...filterArr];
  },

  [actions.actionHandelModal]: (state) => {
    state.isModal = !state.isModal;
  },
  [actions.actionClearCart]: (state) => {
    state.cart = [];
    state.totalPrice = 0;
  },
});
