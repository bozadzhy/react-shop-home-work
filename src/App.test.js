import { render } from "@testing-library/react";
import App from "./App";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import NewThemeProvider from "./context/context";
import store from "./store";

describe("App", () => {
  test("snapshot App", () => {
    const app = render(
      <NewThemeProvider>
        <BrowserRouter>
          <Provider store={store}>
            <App />
          </Provider>
        </BrowserRouter>
      </NewThemeProvider>
    );
    expect(app).toMatchSnapshot();
  });
});
