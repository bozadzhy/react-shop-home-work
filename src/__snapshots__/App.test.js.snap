// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot App 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root"
      >
        <div
          class="container"
        >
          <div>
            <a
              href="/"
            >
              <svg
                class="logo"
              >
                logo.svg
              </svg>
            </a>
          </div>
          <div
            class="wrap"
          >
            <span
              class="theme"
            >
              <svg>
                square1.svg
              </svg>
            </span>
            <a
              class="header-icon"
              href="/favorites"
            >
              <svg>
                starIcon.svg
              </svg>
              <span>
                0
              </span>
            </a>
            <a
              class="header-icon"
              href="/cart"
            >
              <svg>
                cart.svg
              </svg>
              <span>
                0
              </span>
            </a>
          </div>
        </div>
      </div>
      <div
        class="container-list"
      />
      <div
        class="footer"
      >
        Footer
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root"
    >
      <div
        class="container"
      >
        <div>
          <a
            href="/"
          >
            <svg
              class="logo"
            >
              logo.svg
            </svg>
          </a>
        </div>
        <div
          class="wrap"
        >
          <span
            class="theme"
          >
            <svg>
              square1.svg
            </svg>
          </span>
          <a
            class="header-icon"
            href="/favorites"
          >
            <svg>
              starIcon.svg
            </svg>
            <span>
              0
            </span>
          </a>
          <a
            class="header-icon"
            href="/cart"
          >
            <svg>
              cart.svg
            </svg>
            <span>
              0
            </span>
          </a>
        </div>
      </div>
    </div>
    <div
      class="container-list"
    />
    <div
      class="footer"
    >
      Footer
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
